File: study/ts/script/scene/view/cakeActivity/CakeRankNode.ts
import { CakeActivityConst } from "../../../const/CakeActivityConst";
import { handler } from "../../../utils/handler";
import CakeGuildIcon from "./CakeGuildIcon";
import CakePlayerIcon from "./CakePlayerIcon";
import { CakeActivityDataHelper } from "../../../utils/data/CakeActivityDataHelper";
import { Colors, G_UserData } from "../../../init";
import { Path } from "../../../utils/Path";
import UIHelper from "../../../utils/UIHelper";
import { Lang } from "../../../lang/Lang";
import { Util } from "../../../utils/Util";
import CakeRankGuildCell from "./CakeRankGuildCell";
import CakeRankPlayerCell from "./CakeRankPlayerCell";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CakeRankNode extends cc.Component {
    _target: cc.Node;
    _selectTabIndex: number;
    _showArrowPanel: boolean;
    _guildRankTop: {};
    _guildRankList: any[];
    _userRankList: any[];
    _userRankTop: {};

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageTabGuild: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageTabUser: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textGuild: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textUser: cc.Label = null;

    @property({
        type: cc.Button,
        visible: true
    })
    _buttonGuild: cc.Button = null;

    @property({
        type: cc.Button,
        visible: true
    })
    _buttonUser: cc.Button = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeGuild: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodePlayer: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeGuildIcon: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodePlayerIcon: cc.Node = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textGuildEmpty: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textPlayerEmpty: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBgGuild: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBgPlayer: cc.Sprite = null;

    @property({
        type: cc.ScrollView,
        visible: true
    })
    _listViewGuild: cc.ScrollView = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageGuildTip: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textGuildRank: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textGuildScore: cc.Label = null;

    @property({
        type: cc.ScrollView,
        visible: true
    })
    _listViewPlayer: cc.ScrollView = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textMyRank: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textLevelTitle: cc.Label = null;

    

    @property({
        type: cc.Label,
        visible: true
    })
    _textMyScore: cc.Label = null;

    @property({
        type: cc.Button,
        visible: true
    })
    _buttonArrow: cc.Button = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageArrow: cc.Sprite = null;

    @property({
        type: CakeGuildIcon,
        visible: true
    })
    _guildIcon1: CakeGuildIcon = null;

    @property({
        type: CakeGuildIcon,
        visible: true
    })
    _guildIcon2: CakeGuildIcon = null;

    @property({
        type: CakeGuildIcon,
        visible: true
    })
    _guildIcon3: CakeGuildIcon = null;

    @property({
        type: CakePlayerIcon,
        visible: true
    })
    _playerIcon1: CakePlayerIcon = null;

    @property({
        type: CakePlayerIcon,
        visible: true
    })
    _playerIcon2: CakePlayerIcon = null;
    @property({
        type: CakePlayerIcon,
        visible: true
    })
    _playerIcon3: CakePlayerIcon = null;

    onLoad(){
        this.initData();
    }
    initData() {
        this._selectTabIndex = CakeActivityConst.RANK_TYPE_1;
        this._showArrowPanel = false;
        this._guildRankTop = {};
        this._guildRankList = [];
        this._userRankTop = {};
        this._userRankList = [];
        this._initUI();
        this._switchTab();
        this._updateInfoPanel();
    }
    _initUI() {
        for (var i = 1; i <= 3; i++) {
            this['_guildIcon' + i].initData(i);
            this['_playerIcon' + i].ctor(i);
        }

        var name = CakeActivityDataHelper.getFoodName();
        this._textLevelTitle.string = (Lang.get('cake_activity_rank_level_title', { name: name }));
        this.updateStage();
    }
    onClickTabGuild() {
        if (this._selectTabIndex == CakeActivityConst.RANK_TYPE_1) {
            return;
        }
        this._selectTabIndex = CakeActivityConst.RANK_TYPE_1;
        this._switchTab();
    }
    onClickTabUser() {
        if (this._selectTabIndex == CakeActivityConst.RANK_TYPE_2) {
            return;
        }
        this._selectTabIndex = CakeActivityConst.RANK_TYPE_2;
        this._switchTab();
    }
    updateStage() {
        var actStage = CakeActivityDataHelper.getActStage()[0];
        if (actStage == CakeActivityConst.ACT_STAGE_1 || actStage == CakeActivityConst.ACT_STAGE_2) {
            this._textGuild.string = (Lang.get('cake_activity_rank_tab_title_1'));
            this._textUser.string = (Lang.get('cake_activity_rank_tab_title_2'));
            this._imageGuildTip.node.active = (true);
        } else {
            this._textGuild.string = (Lang.get('cake_activity_rank_tab_title_3'));
            this._textUser.string = (Lang.get('cake_activity_rank_tab_title_4'));
            this._imageGuildTip.node.active = (false);
        }
    }
    _switchTab() {
        this._imageTabGuild.node.active = (this._selectTabIndex == CakeActivityConst.RANK_TYPE_1);
        this._imageTabUser.node.active = (this._selectTabIndex == CakeActivityConst.RANK_TYPE_2);
        this._nodeGuild.active = (this._selectTabIndex == CakeActivityConst.RANK_TYPE_1);
        this._nodePlayer.active = (this._selectTabIndex == CakeActivityConst.RANK_TYPE_2);
        if (this._selectTabIndex == CakeActivityConst.RANK_TYPE_1) {
            this._textGuild.node.color = (new cc.Color(200, 94, 9));
            this._textUser.node.color = (new cc.Color(237, 153, 91));
        } else {
            this._textGuild.node.color = (new cc.Color(237, 153, 91));
            this._textUser.node.color = (new cc.Color(200, 94, 9));
        }
        this.updateRank();
    }
    onClickArrow() {
        this._showArrowPanel = !this._showArrowPanel;
        this._updateInfoPanel();
    }
    _updateInfoPanel() {
        if (this._showArrowPanel) {
            UIHelper.loadTexture(this._imageArrow, Path.getTextAnniversaryImg('txt_anniversary01', null));
        } else {
            UIHelper.loadTexture(this._imageArrow, Path.getTextAnniversaryImg('txt_anniversary02', null));
        }
        this._imageBgGuild.node.active = (this._showArrowPanel);
        this._imageBgPlayer.node.active = (this._showArrowPanel);
    }
    updateRank() {
        if (this._selectTabIndex == CakeActivityConst.RANK_TYPE_1) {
            this.updateGuildRank();
        } else if (this._selectTabIndex == CakeActivityConst.RANK_TYPE_2) {
            this.updatePlayerRank();
            this.updateMyScore();
        }
    }
    updateGuildRank() {
        var rankData = G_UserData.getCakeActivity().getSelfGuildRank();
        if (rankData) {
            this._textGuildRank.string = (rankData.getRank());
            this._textGuildScore.string = (rankData.getCake_level());
        } else {
            this._textGuildRank.string = (Lang.get('cake_activity_no_rank'));
            this._textGuildScore.string = (Lang.get('cake_activity_no_level'));
        }
        this._guildRankTop = {};
        this._guildRankList = [];
        var listDatas = G_UserData.getCakeActivity().getGuildRankList();
        if (listDatas.length == 0) {
            this._nodeGuildIcon.active = (false);
            this._textGuildEmpty.node.active = (true);
        } else {
            this._nodeGuildIcon.active = (true);
            this._textGuildEmpty.node.active = (false);
        }
        for (var i = 0; i < 3; i++) {
            this._guildRankTop[i] = listDatas[i];
        }
        for (var i = 3; i < listDatas.length; i++) {
            this._guildRankList.push(listDatas[i]);
        }
        for (var i = 1; i <= 3; i++) {
            var data = this._guildRankTop[i - 1];
            this['_guildIcon' + i].updateUI(data);
        }
        this._listViewGuild.content.removeAllChildren();
        this._listViewGuild.content.height = 334;
        for (let i = 0; i < this._guildRankList.length; i++) {
            let cell = Util.getNode("prefab/cakeActivity/CakeRankGuildCell", CakeRankGuildCell) as CakeRankGuildCell;
            this._listViewGuild.content.addChild(cell.node);
            cell.updateUI(this._guildRankList[i]);
            cell.node.x = 0;
            cell.node.y = (-i - 1) * 40;
            if (Math.abs(cell.node.y) > 334) {
                this._listViewGuild.content.height = Math.abs(cell.node.y);
            }
        }
    }
    updatePlayerRank() {
        var rankData = G_UserData.getCakeActivity().getSelfUserRank();
        if (rankData) {
            this._textMyRank.string = (rankData.getRank());
        } else {
            this._textMyRank.string = (Lang.get('cake_activity_no_rank'));
        }
        this._userRankTop = {};
        this._userRankList = [];
        var listDatas = G_UserData.getCakeActivity().getUserRankList();
        if (listDatas.length == 0) {
            this._nodePlayerIcon.active = (false);
            this._textPlayerEmpty.node.active = (true);
        } else {
            this._nodePlayerIcon.active = (true);
            this._textPlayerEmpty.node.active = (false);
        }
        for (var i = 0; i < 3; i++) {
            this._userRankTop[i] = listDatas[i];
        }
        for (var i = 3; i < listDatas.length; i++) {
            this._userRankList.push(listDatas[i]);
        }
        for (var i = 1; i <= 3; i++) {
            var data = this._userRankTop[i - 1];
            this['_playerIcon' + i].updateUI(data);
        }
        this._listViewPlayer.content.removeAllChildren();
        this._listViewPlayer.content.height = 360;
        for (let i = 0; i < this._userRankList.length; i++) {
            let cell = Util.getNode("prefab/cakeActivity/CakeRankPlayerCell", CakeRankPlayerCell) as CakeRankPlayerCell;
            this._listViewPlayer.content.addChild(cell.node);
            cell.updateUI(this._userRankList[i]);
            cell.node.x = 0;
            cell.node.y = (-i - 1) * 40;
            if (Math.abs(cell.node.y) > 360) {
                this._listViewPlayer.content.height = Math.abs(cell.node.y);
            }
        }
    }
    updateMyScore() {
        var point = G_UserData.getCakeActivity().getSelfPoint();
        this._textMyScore.string = (point);
    }
    _onItemUpdate1(item, index) {
        var index = index + 1;
        var data = this._guildRankList[index];
        if (data) {
            item.update(data);
        }
    }
    _onItemSelected1(item, index) {
    }
    _onItemTouch1(index, t) {
    }
    _onItemUpdate2(item, index) {
        var index = index + 1;
        var data = this._userRankList[index];
        if (data) {
            item.update(data);
        }
    }
    _onItemSelected2(item, index) {
    }
    _onItemTouch2(index, t) {
    }

}

File: study/lua/src/app/scene/view/cakeActivity/CakeRankNode.lua
--
-- Author: Liangxu
-- Date: 2019-4-30
-- 蛋糕活动排行榜

local CakeRankNode = class("CakeRankNode")
local CakeGuildIcon = require("app.scene.view.cakeActivity.CakeGuildIcon")
local CakePlayerIcon = require("app.scene.view.cakeActivity.CakePlayerIcon")
local CakeActivityConst = require("app.const.CakeActivityConst")
local CakeRankGuildCell = require("app.scene.view.cakeActivity.CakeRankGuildCell")
local CakeRankPlayerCell = require("app.scene.view.cakeActivity.CakeRankPlayerCell")
local CakeActivityDataHelper = require("app.utils.data.CakeActivityDataHelper")

function CakeRankNode:ctor(target)
	self._target = target
	self._selectTabIndex = CakeActivityConst.RANK_TYPE_1

	self._showArrowPanel = false --默认关闭
	self._guildRankTop = {} --前3名
	self._guildRankList = {} --3名之后
	self._userRankTop = {} --前3名
	self._userRankList = {} --3名之后

	self:_initUI()
	self:_switchTab()
	self:_updateInfoPanel()
end

function CakeRankNode:_initUI()
	self._imageTabGuild = ccui.Helper:seekNodeByName(self._target, "ImageTabGuild")
	self._imageTabUser = ccui.Helper:seekNodeByName(self._target, "ImageTabUser")
	self._textGuild = ccui.Helper:seekNodeByName(self._target, "TextGuild")
	self._textUser = ccui.Helper:seekNodeByName(self._target, "TextUser")
	self._buttonGuild = ccui.Helper:seekNodeByName(self._target, "ButtonGuild")
	self._buttonGuild:addClickEventListenerEx(handler(self, self._onClickTabGuild))
	self._buttonUser = ccui.Helper:seekNodeByName(self._target, "ButtonUser")
	self._buttonUser:addClickEventListenerEx(handler(self, self._onClickTabUser))

	self._nodeGuild = ccui.Helper:seekNodeByName(self._target, "NodeGuild")
	self._nodePlayer = ccui.Helper:seekNodeByName(self._target, "NodePlayer")
	self._nodeGuildIcon = ccui.Helper:seekNodeByName(self._target, "NodeGuildIcon")
	self._nodePlayerIcon = ccui.Helper:seekNodeByName(self._target, "NodePlayerIcon")
	for i = 1, 3 do
		local nodeGuildIcon = ccui.Helper:seekNodeByName(self._target, "NodeGuildIcon"..i)
		self["_guildIcon"..i] = CakeGuildIcon.new(nodeGuildIcon, i)
		local nodePlayerIcon = ccui.Helper:seekNodeByName(self._target, "NodePlayerIcon"..i)
		self["_playerIcon"..i] = CakePlayerIcon.new(nodePlayerIcon, i) 
	end
	self._textGuildEmpty = ccui.Helper:seekNodeByName(self._target, "TextGuildEmpty")
	self._textPlayerEmpty = ccui.Helper:seekNodeByName(self._target, "TextPlayerEmpty")
	self._imageBgGuild = ccui.Helper:seekNodeByName(self._target, "ImageBgGuild")
	self._imageBgPlayer = ccui.Helper:seekNodeByName(self._target, "ImageBgPlayer")

	self._listViewGuild = ccui.Helper:seekNodeByName(self._target, "ListViewGuild")
	cc.bind(self._listViewGuild, "ScrollView")
	self._listViewGuild:setTemplate(CakeRankGuildCell)
	self._listViewGuild:setCallback(handler(self, self._onItemUpdate1), handler(self, self._onItemSelected1))
	self._listViewGuild:setCustomCallback(handler(self, self._onItemTouch1))
	self._imageGuildTip = ccui.Helper:seekNodeByName(self._target, "ImageGuildTip")
	self._textGuildRank = ccui.Helper:seekNodeByName(self._target, "TextGuildRank")
	self._textGuildScore = ccui.Helper:seekNodeByName(self._target, "TextGuildScore")
	-- self._imageGuildArrow = ccui.Helper:seekNodeByName(self._target, "ImageGuildArrow")
	-- self._imageGuildArrow:ignoreContentAdaptWithSize(true)

	self._listViewPlayer = ccui.Helper:seekNodeByName(self._target, "ListViewPlayer")
	cc.bind(self._listViewPlayer, "ScrollView")
	self._listViewPlayer:setTemplate(CakeRankPlayerCell)
	self._listViewPlayer:setCallback(handler(self, self._onItemUpdate2), handler(self, self._onItemSelected2))
	self._listViewPlayer:setCustomCallback(handler(self, self._onItemTouch2))
	self._textMyRank = ccui.Helper:seekNodeByName(self._target, "TextMyRank")
	self._textMyScore = ccui.Helper:seekNodeByName(self._target, "TextMyScore")
	-- self._imageMyArrow = ccui.Helper:seekNodeByName(self._target, "ImageMyArrow")
	-- self._imageMyArrow:ignoreContentAdaptWithSize(true)

	self._buttonArrow = ccui.Helper:seekNodeByName(self._target, "ButtonArrow")
	self._buttonArrow:addClickEventListenerEx(handler(self, self._onClickArrow))
	self._imageArrow = ccui.Helper:seekNodeByName(self._target, "ImageArrow")

	self._textLevelTitle = ccui.Helper:seekNodeByName(self._target, "TextLevelTitle")
	local name = CakeActivityDataHelper.getFoodName()
	self._textLevelTitle:setString(Lang.get("cake_activity_rank_level_title", {name = name}))

	self:updateStage()
end

function CakeRankNode:_onClickTabGuild()
	if self._selectTabIndex == CakeActivityConst.RANK_TYPE_1 then
		return
	end
	self._selectTabIndex = CakeActivityConst.RANK_TYPE_1
	self:_switchTab()
end

function CakeRankNode:_onClickTabUser()
	if self._selectTabIndex == CakeActivityConst.RANK_TYPE_2 then
		return
	end
	self._selectTabIndex = CakeActivityConst.RANK_TYPE_2
	self:_switchTab()
end

function CakeRankNode:updateStage()
	local actStage = CakeActivityDataHelper.getActStage()
	if actStage == CakeActivityConst.ACT_STAGE_1 or actStage == CakeActivityConst.ACT_STAGE_2 then
		self._textGuild:setString(Lang.get("cake_activity_rank_tab_title_1"))
		self._textUser:setString(Lang.get("cake_activity_rank_tab_title_2"))
		self._imageGuildTip:setVisible(true)
	else
		self._textGuild:setString(Lang.get("cake_activity_rank_tab_title_3"))
		self._textUser:setString(Lang.get("cake_activity_rank_tab_title_4"))
		self._imageGuildTip:setVisible(false)
	end
end

function CakeRankNode:_switchTab()
	self._imageTabGuild:setVisible(self._selectTabIndex == CakeActivityConst.RANK_TYPE_1)
	self._imageTabUser:setVisible(self._selectTabIndex == CakeActivityConst.RANK_TYPE_2)
	self._nodeGuild:setVisible(self._selectTabIndex == CakeActivityConst.RANK_TYPE_1)
	self._nodePlayer:setVisible(self._selectTabIndex == CakeActivityConst.RANK_TYPE_2)
	if self._selectTabIndex == CakeActivityConst.RANK_TYPE_1 then
		self._textGuild:setColor(cc.c3b(0xc8, 0x5e, 0x09))
		self._textUser:setColor(cc.c3b(0xed, 0x99, 0x5b))
	else
		self._textGuild:setColor(cc.c3b(0xed, 0x99, 0x5b))
		self._textUser:setColor(cc.c3b(0xc8, 0x5e, 0x09))
	end
	self:updateRank()
end

function CakeRankNode:_onClickArrow()
	self._showArrowPanel = not self._showArrowPanel
	self:_updateInfoPanel()
end

function CakeRankNode:_updateInfoPanel()
	if self._showArrowPanel then
		self._imageArrow:loadTexture(Path.getTextAnniversaryImg("txt_anniversary01"))
	else
		self._imageArrow:loadTexture(Path.getTextAnniversaryImg("txt_anniversary02"))
	end
	self._imageBgGuild:setVisible(self._showArrowPanel)
	self._imageBgPlayer:setVisible(self._showArrowPanel)
end

function CakeRankNode:updateRank()
	if self._selectTabIndex == CakeActivityConst.RANK_TYPE_1 then
		self:updateGuildRank()
	elseif self._selectTabIndex == CakeActivityConst.RANK_TYPE_2 then
		self:updatePlayerRank()
		self:updateMyScore()
	end
end

function CakeRankNode:updateGuildRank()
	local rankData = G_UserData:getCakeActivity():getSelfGuildRank()
	if rankData then
		self._textGuildRank:setString(rankData:getRank())
		self._textGuildScore:setString(rankData:getCake_level())
		-- self._imageGuildArrow:loadTexture(Path.getUICommon(rankData:getChangeResName()))
		-- self._imageGuildArrow:setVisible(true)
	else
		self._textGuildRank:setString(Lang.get("cake_activity_no_rank"))
		self._textGuildScore:setString(Lang.get("cake_activity_no_level"))
		-- self._imageGuildArrow:setVisible(false)
	end

	self._guildRankTop = {}
	self._guildRankList = {}
	local listDatas = G_UserData:getCakeActivity():getGuildRankList()
	if #listDatas == 0 then
		self._nodeGuildIcon:setVisible(false)
		self._textGuildEmpty:setVisible(true)
	else
		self._nodeGuildIcon:setVisible(true)
		self._textGuildEmpty:setVisible(false)
	end
	
	for i = 1, 3 do
		self._guildRankTop[i] = listDatas[i]
	end
	for i = 4, #listDatas do
		table.insert(self._guildRankList, listDatas[i])
	end

	for i = 1, 3 do
		local data = self._guildRankTop[i]
		self["_guildIcon"..i]:updateUI(data)
	end
	self._listViewGuild:clearAll()
	self._listViewGuild:resize(#self._guildRankList)
end

function CakeRankNode:updatePlayerRank()
	local rankData = G_UserData:getCakeActivity():getSelfUserRank()
	if rankData then
		self._textMyRank:setString(rankData:getRank())
		-- self._imageMyArrow:loadTexture(Path.getUICommon(rankData:getChangeResName()))
		-- self._imageMyArrow:setVisible(true)
	else
		self._textMyRank:setString(Lang.get("cake_activity_no_rank"))
		-- self._imageMyArrow:setVisible(false)
	end

	self._userRankTop = {}
	self._userRankList = {}
	local listDatas = G_UserData:getCakeActivity():getUserRankList()
	if #listDatas == 0 then
		self._nodePlayerIcon:setVisible(false)
		self._textPlayerEmpty:setVisible(true)
	else
		self._nodePlayerIcon:setVisible(true)
		self._textPlayerEmpty:setVisible(false)
	end
	
	for i = 1, 3 do
		self._userRankTop[i] = listDatas[i]
	end
	for i = 4, #listDatas do
		table.insert(self._userRankList, listDatas[i])
	end

	for i = 1, 3 do
		local data = self._userRankTop[i]
		self["_playerIcon"..i]:updateUI(data)
	end
	self._listViewPlayer:clearAll()
	self._listViewPlayer:resize(#self._userRankList)
end

function CakeRankNode:updateMyScore()
	local point = G_UserData:getCakeActivity():getSelfPoint()
	self._textMyScore:setString(point)
end

function CakeRankNode:_onItemUpdate1(item, index)
    local index = index + 1
    local data = self._guildRankList[index]
    if data then
    	item:update(data)
    end
end

function CakeRankNode:_onItemSelected1(item, index)
	
end

function CakeRankNode:_onItemTouch1(index, t)

end

function CakeRankNode:_onItemUpdate2(item, index)
    local index = index + 1
    local data = self._userRankList[index]
    if data then
    	item:update(data)
    end
end

function CakeRankNode:_onItemSelected2(item, index)
	
end

function CakeRankNode:_onItemTouch2(index, t)

end

return CakeRankNode